### ISSUE TITLE
Remove-PSDrive does not work correctly for Networked Drives created using -persist and -scope Global

### ISSUE THREAD

**Issue by VortexUK** [PowerShell/PowerShell#7829]
To be clear, all the above was done in PowerShell Core, however I can also reproduce this in PoSH. Almost Identical to #7193 

Steps to reproduce
------------------
2 separate sessions required

Session 1
```
New-PSDrive -Name J -PSProvider FileSystem -Root "\\contoso.com\share" -Description Test -Scope Global -Persist
```

Session 2
```
> get-psdrive -name J

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
J                   2.44      10237.44 FileSystem    \\contoso.com\share


> remove-psdrive -name J
> get-psdrive -name J

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
J                   2.44      10237.44 FileSystem    \\contoso.com\share

```

Expected behavior
-----------------
For it to actually delete J, similar to 'net use J: /delete'
```none

```

Actual behavior
---------------
It does not delete J, or error. running
```none
PS > net use J: /delete
J: was deleted successfully.
```
Works as expected

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.0
PSEdition                      Core
GitCommitId                    6.1.0
OS                             Microsoft Windows 10.0.14393
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by vexx32** [comment]
Does it work using `Remove-PSDrive -Scope Global`?

**Comment by MaximoTrinidad** [comment]
@VortexUK,

Greatly Appreciate submitting this issue.  
:)

**Comment by thezim** [comment]
To be clear *-PSDrive cmdlets are not using the same API as `net use`. You can't cross that boundary are far as I'm aware.

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/new-psdrive?view=powershell-6

**Comment by VortexUK** [comment]
I wasn't suggesting they are the same @thezim , it was more showing that it was *possible* to delete the drive using other tools.

@vexx32 - same issue:

```
PS C:\Program Files\PowerShell\6.1.0-preview> get-psdrive -name J

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
J                   2.44      10237.44 FileSystem    \\contoso.com\share


PS C:\Program Files\PowerShell\6.1.0-preview>  remove-psdrive -name J -Scope Global
PS C:\Program Files\PowerShell\6.1.0-preview> get-psdrive -name J

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
J                   2.44      10237.44 FileSystem    \\contoso.com\share
```

**Comment by agoldenlife** [comment]
For those finding this.  I went looking and found someone much smarter than I who knows how to do this properly.  There are several ways to create network drives and Remove-PSDrive is not a catch all to remove them.

Source
https://blogs.msmvps.com/russel/tag/remove-psdrive/

Here is the code in case the link ever breaks
```
<#
.SYNOPSIS
Unmaps network drives
.DESCRIPTION
Unmapdrives removes all currently mapped network drives. It's smart enough to 
remove drives mapped with "net use", "New-SmbMapping" and "New-PSDrive". This 
cmdlet accepts no parameters and assumes -Force for all unmappings. 

.EXAMPLE
UnMapDrives 
Unmaps all currently mapped network drives 

.NOTES
    Author: Charlie Russel
 Copyright: 2015 by Charlie Russel
          : Permission to use is granted but attribution is appreciated
   Initial: 06/27/2015 (cpr)
   ModHist:
 :
#>
[CmdletBinding()]

# Build a dynamic list of currently mapped drives
$DriveList = Get-WMIObject Win32_LogicalDisk `
     | Where-Object { $_.DriveType -eq 4 }

# Don't bother running this if we don't have any mapped drives
 if ($DriveList) { 
    $SmbDriveList = $DriveList.DeviceID
 } else {
    Write-Host "No mapped drives found"
    Return
}

Write-host "Unmapping drive: " -NoNewLine
Write-Host $SmbDriveList
Write-Host " "

Foreach ($drive in $SmbDriveList) {
    $psDrive = $drive -replace ":" #remove unwanted colon from PSDrive name
    Remove-SmbMapping -LocalPath $Drive -Force -UpdateProfile
    If ( (Get-PSDrive -Name $psDrive) 2>$Null ) {
       Remove-PSDrive -Name $psDrive -Force
    }
}
Write-Host " "

# Report back all FileSystem drives to confirm that only local drives are present. 
Get-PSDrive -PSProvider FileSystem`
```

**Comment by rusticMire** [comment]
Any updates on this issue? It's affecting my build pipelines. I have to use a `New-PSDrive` to add and `net use` to remove Network Drives. This inconsistency in my pipeline doesn't look great.

**Comment by mkanet** [comment]
@agoldenlife  Thanks for the script.  But geez. I think it's a lot less hassle to just add the extra line below in (addition to Remove-PSDrive. if necessary) to accomplish removing the drive completely.  Pretty lame that they're interfering with one another.

NET USE /del \\$Device 2>null

**Comment by agoldenlife** [comment]
@mkanet I think you miss the point.  

In the code above you simply type "unMapDrives" and all of them are now removed.  I don't see how typing two commands (i.e. Remove-PSDrive and then net use /del \$Device 2>null) is any easier.

The above code is written you so can load it into your powershell profile.   Once that is set you can use it as much as you want.  From then on out you just type the one command.  It is also nice because it gives you feedback and makes checks to ensure it all worked properly.  Your script outputs any errors to null so you have to check manually if it was successful or not.  Of course that may be what you want so... ymmv. 

The nice thing about scripting is there are many ways to accomplish the same thing!  Thanks for commenting and showing what works for you.

**Comment by mkanet** [comment]
@agoldenlife, I wasn't referring to the script you posted.  I meant using native PowerShell and ad-hoc cases; referring to to the OP of this thread.  Creating an UnMapDrives script, loading it, then execute the command would take significantly longer than just typing a couple of commands.  But yeah, if unMapDrives is already loaded in memory, it would be even quicker.

**Comment by erict-powersoft** [comment]
Still broken on Windows 10  Enterprise 1809 build 17763.615

**Comment by MaximoTrinidad** [comment]
Hum! Although, it's broken in Windows PowerShell (no more changes here), and it hasn't been fix in PowerShell 7 (Core), there's a workaround.

Either install the SMB components in the client machine, which seems to work flawlessly, or use the following command line:
```
Get-PSdrive -Name t | Remove-PSDrive
```
Now, this one-liner will give an error:
```
ps> Get-PSdrive -Name t | Remove-PSDrive
Remove-PSDrive : Cannot find drive. A drive with the name 't' does not exist.
At line:1 char:23
+ Get-PSDrive -Name t | Remove-PSDrive
+                       ~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (t:String) [Remove-PSDrive], DriveNotFoundException
    + FullyQualifiedErrorId : DriveNotFound,Microsoft.PowerShell.Commands.RemovePSDriveCommand

Remove-PSDrive : Cannot find drive. A drive with the name 't' does not exist.
At line:1 char:23
+ Get-PSdrive -Name t | Remove-PSDrive
+                       ~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (t:String) [Remove-PSDrive], DriveNotFoundException
    + FullyQualifiedErrorId : DriveNotFound,Microsoft.PowerShell.Commands.RemovePSDriveCommand
```
![GetPSDrive_01_2019-08-15_18-42-53](https://user-images.githubusercontent.com/19892435/63131846-1f6c5c80-bf8d-11e9-835f-790ed700959f.png)

But, even with the error message, the drive is removed. 

So, manually handle the error by including both the ErrorAction and ErrorVariable parameter to prevent the error to be displayed.
```
Get-PSdrive -Name t | Remove-PSDrive -ErrorAction SilentlyContinue -ErrorVariable rmDrive
```
![GetPSDrive_02_2019-08-15_18-42-53](https://user-images.githubusercontent.com/19892435/63132267-82122800-bf8e-11e9-8cf4-75fcb7154106.png)

I know is just a workaround, and it does work in both versions of PowerShell (Windows and Core).

We'll have to wait for what the PowerShell Team decide to do about this one with PowerShell 7.

:)

**Comment by dendy7h** [comment]
Have you tried using the force? I had the same issue and found the answer in agoldenlife's script:
```Remove-PSDrive -Name <drive_name> -Force```

**Comment by MaximoTrinidad** [comment]
@dendy7h 

-Force parameter is not implemented in this cmdlet. :(

**Comment by MaximoTrinidad** [comment]
Just an FYI.

This works in PowerShell 7 Preview RC2, even without using the -Force parameter.

![PS7_removeDrive_2020-02-19_11-52-18](https://user-images.githubusercontent.com/19892435/74857234-69097b80-5311-11ea-977c-6c17cdfd4506.png)

And, it also works in PowerShell 6.2.4:
![PS624_removeDrive_2020-02-19_11-52-18](https://user-images.githubusercontent.com/19892435/74857547-d1585d00-5311-11ea-8756-b58e2db25b58.png)

**Comment by ScriptingPro** [comment]
None of this working in PowerShell 5.1
![image](https://user-images.githubusercontent.com/13045541/76897434-bf5ed100-6861-11ea-9d4a-5329e197f763.png)

**Comment by MaximoTrinidad** [comment]
@ScriptingPro 

Keep in mind! Windows PowerShell 5.1 is complete. So, any fixes are done against PowerShell 7 (or Greater). 

No fixes will be done to Windows PowerShell 5.1.  Only security updates.
:)

**Comment by vexx32** [comment]
Aye. If it's already fixed in PS 7, we can close this one. 🙂

**Comment by justinsantoro** [comment]
I don't think this is actually fixed. Like @MaximoTrinidad I am able to remove the drive if I had created it _within the same session_. However, like OP, If I create the persisted drive in one session then try to remove it in _another_ session, it does not work.

Am I missing something?

```Powershell
> PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.19041
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by MaximoTrinidad** [comment]
@justinsantoro 

I think it make sense that if the drive was created, and persisted, in session 1, that session 2 won't be able to remove it as it is still in use by session 1. 

:)

**Comment by vexx32** [comment]
If the first session is closed by the time session 2 tries to remove it, it should still be removable I would think. Is that the case @justinsantoro?

**Comment by simeyla** [comment]
What a nightmare! (as Windows networking always seems to be)

You can also just right click on the drive in Explorer and click `Disconnect` :-) 
You may need to reopen explorer.

**Comment by elliot-labs** [comment]
I have just reproduced this issue with PowerShell 7.0.0.

**Comment by VortexUK** [comment]
> @justinsantoro
> 
> I think it make sense that if the drive was created, and persisted, in session 1, that session 2 won't be able to remove it as it is still in use by session 1.
> 
> :)

That doesn't make sense at all. The drive is a _persistent_ mapped drive, meaning it will continue well after a session has gone, through log outs and reboots. Conversely a session is exactly that - something temporary. The drive should be un-mappable from any session you chose with the right command, but, as shown in the original post, it is not. Also I should have mentioned in the original post, but -force, -scope etc have no affect on this bug. It... persists.


> I have just reproduced this issue with PowerShell 7.0.0.

This is really disappointing!

**Comment by elliot-labs** [comment]
@vexx32 , can you please reopen this issue?
This is not fixed in 7.0.0

**Comment by TheBlaide** [comment]
The issue still persists in pws 7.0.1.

**Comment by MaximoTrinidad** [comment]
@TheBlaide 

Hum! It works for me! No issues at all.

![PSDriveWorks_01_2020-06-05_10-28-14](https://user-images.githubusercontent.com/19892435/83890382-b58a9280-a719-11ea-85f6-12f77caf19c4.jpg)

![PSDriveWorks_02_2020-06-05_10-28-14](https://user-images.githubusercontent.com/19892435/83890381-b4f1fc00-a719-11ea-9a19-42a04e27ffe1.jpg)

:)

**Comment by vexx32** [comment]
Reopening as some folks are still seeing this issue. Can we get environment data and exactly what code y'all are running to get the errors you're seeing as well, please? 🙂

**Comment by MaximoTrinidad** [comment]
@vexx32 

Sure! Here's my code to repo:

```
## - Create new mapped drive "W":
New-PSDrive -Name "W" -psprovider FileSystem `
	-root "\\SUN\MyWinSvrSharedFiles" -persist `
	-description "Server Shared folder" `
	-Credential (Get-Credential);

## - Display all Filesystem Drives including the new mapped drive:
Get-PSDrive -PSProvider FileSystem;

## - Remove mapped drive "W":
Get-PSDrive -Name "W" | Remove-PSDrive;

## - Display all Filesystem Drives but mapped drive was removed:
Get-PSDrive -PSProvider FileSystem;

```
:)

**Comment by vexx32** [comment]
@MaximoTrinidad it looks like you're not using `-Scope Global` when you create that drive?

**Comment by VortexUK** [comment]
@MaximoTrinidad didn't read the original problem and created/deleted the drive in the same session. This is still an issue.

**Comment by vexx32** [comment]
Seems to be the case. @VortexUK out of curiosity does the same reproduce without `-Scope Global` or is it specifically an issue with that switch in combination with the others?

**Comment by VortexUK** [comment]
It's been a while since I retested - I'll see if I can do it with the latest release I have and  get back to you

**Comment by VortexUK** [comment]
Just confirmed on 7.0.0 (OS Windows 10.0.17134)

`New-PSDrive -Name J -PSProvider FileSystem -Root \\contoso.com\someshare -Scope Global -Persist`

Works as intended. When you then open another powershell session, you can see it doing

`Get-PSDrive -Name J`

However, the following have no affect, the drive remains:

```
Get-PSDrive -Name J | Remove-PSDrive
Get-PSDrive -Name J | Remove-PSDrive -Force
```

Going back to the original session and running the above will work.

**Comment by MaximoTrinidad** [comment]
@VortexUK 

Just in case!
Can you try with the latest PowerShell GA is version 7.0.1?

:)

**Comment by szymonos** [comment]
PowerShell 7.1.1 and the problem still exists. The only way to remove persisted drive is to use "net use X: /delete"

**Comment by kborowinski** [comment]
> Just confirmed on 7.0.0 (OS Windows 10.0.17134)
> 
> `New-PSDrive -Name J -PSProvider FileSystem -Root \\contoso.com\someshare -Scope Global -Persist`
> 
> Works as intended. When you then open another powershell session, you can see it doing
> 
> `Get-PSDrive -Name J`
> 
> However, the following have no affect, the drive remains:
> 
> ```
> Get-PSDrive -Name J | Remove-PSDrive
> Get-PSDrive -Name J | Remove-PSDrive -Force
> ```
> 
> Going back to the original session and running the above will work.

This is still happening in **PowerShell 7.2.0 Preview 9**. *Remove-PSDrive* cmdlet for *network drives* works only when executed in the same session on which the *New-PSDrive* was run before.

**Comment by AgostinoX** [comment]
A bug that seems to be related to this and maybe it can help to find the solution.
On PS 5.1.19041.1320, OS: Win 10 Home; version 10.0.19043 N/D build 19043
Firts of all: as others said, in the same session it's possible to remove the PSDrive: in the code below, after the Remove-PSDrive the Get-PSDrive doesn't return X anymore.

```
PS C:\Users\Agostino> New-PSDrive -Name X -PSProvider FileSystem \\localhost\share -Persist -Scope global

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
X                  93,08         11,41 FileSystem    \\localhost\share


PS C:\Users\Agostino> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
Alias                                  Alias
C                  93,08         11,41 FileSystem    C:\                                                 Users\Agostino
Cert                                   Certificate   \
Env                                    Environment
Function                               Function
HKCU                                   Registry      HKEY_CURRENT_USER
HKLM                                   Registry      HKEY_LOCAL_MACHINE
Variable                               Variable
WSMan                                  WSMan
X                  93,08         11,41 FileSystem    \\localhost\share


PS C:\Users\Agostino> Remove-PSDrive X
PS C:\Users\Agostino> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
Alias                                  Alias
C                  93,08         11,41 FileSystem    C:\                                                 Users\Agostino
Cert                                   Certificate   \
Env                                    Environment
Function                               Function
HKCU                                   Registry      HKEY_CURRENT_USER
HKLM                                   Registry      HKEY_LOCAL_MACHINE
Variable                               Variable
WSMan                                  WSMan

```

Now let's do something **strange**. I issue the same command-let again, but replacing X with Y (just to be sure to have a 'clean' situation) but using a non-existent scope option ("dummyscope"). 
It obviously gives an error (sorry it's in italian, but it's just the classic invalid-argument error). 
Then something strange happens. The drive gets mapped the same (at system level too) but goes into the very state of non-removable drives that's the issue studied here.
To get this behavior I imagine that the -Scope parameter is not validated at the beginning of the function, but after something (the actual mapping) has already been done but the error sort of prevents the registration of the unit where it should be to get correctly removed after. 

```
PS C:\Users\Agostino> New-PSDrive -Name Y -PSProvider FileSystem \\localhost\share -Persist -Scope dummyscope
New-PSDrive : Non è possibile elaborare l'argomento perché il valore dell'argomento "scopeID" non è valido. I valori
validi sono "Global", "Local", "Script" o un numero relativo all'ambito corrente (da 0 al numero di ambiti, dove 0 è
l'ambito corrente e 1 è il relativo padre). Modifica il valore dell'argomento "scopeID" ed esegui di nuovo
l'operazione.
In riga:1 car:1
+ New-PSDrive -Name Y -PSProvider FileSystem \\localhost\share -Persist ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (Y:PSDriveInfo) [New-PSDrive], PSArgumentException
    + FullyQualifiedErrorId : NewDriveError,Microsoft.PowerShell.Commands.NewPSDriveCommand

PS C:\Users\Agostino> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
Alias                                  Alias
C                  93,08         11,41 FileSystem    C:\                                                 Users\Agostino
Cert                                   Certificate   \
Env                                    Environment
Function                               Function
HKCU                                   Registry      HKEY_CURRENT_USER
HKLM                                   Registry      HKEY_LOCAL_MACHINE
Variable                               Variable
WSMan                                  WSMan
Y                  93,08         11,41 FileSystem    \\localhost\share


PS C:\Users\Agostino> Remove-PSDrive Y
PS C:\Users\Agostino> Get-PSDrive

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
Alias                                  Alias
C                  93,08         11,41 FileSystem    C:\                                                 Users\Agostino
Cert                                   Certificate   \
Env                                    Environment
Function                               Function
HKCU                                   Registry      HKEY_CURRENT_USER
HKLM                                   Registry      HKEY_LOCAL_MACHINE
Variable                               Variable
WSMan                                  WSMan
Y                  93,08         11,41 FileSystem    \\localhost\share


```

**Comment by martens-d** [comment]
I'm on Windows Server 2016 Standard, the PS Version is 5.1.14393.6343.

When I use "-Persist -Scope Global" only than am I able to access the drive from another cmd window. But when I try to remove it I get something very strange:
1.
![grafik](https://github.com/PowerShell/PowerShell/assets/73710614/9526c79d-1646-4c08-ab83-110879af1f0b)
I have tested with A and B as drive letters so there are two in here:
![grafik](https://github.com/PowerShell/PowerShell/assets/73710614/02f3b0cd-4efb-4a82-8f3c-fca6942f5e89)
When I try to remove them with "Remove-PSDrive -name A -Force -Scope Global"
The drive letter is mapped like a local drive: 
![grafik](https://github.com/PowerShell/PowerShell/assets/73710614/432a5513-ccac-4ad5-9a72-05dacc1846e0)

I tried using the portable version of PS but to no avail.

Can anybody help me?

Thanks
